File: scripts/analisar-ocorrencias-antigas.py
import os
import re
import pandas as pd
from pathlib import Path
from dotenv import load_dotenv
from supabase import create_client

load_dotenv(Path.cwd() / '.env')

url = os.getenv('VITE_SUPABASE_URL')
key = os.getenv('SUPABASE_SERVICE_ROLE_KEY') or os.getenv('VITE_SUPABASE_ANON_KEY')

if not url or not key:
    raise ValueError('Variáveis VITE_SUPABASE_URL e SUPABASE_SERVICE_ROLE_KEY são obrigatórias')

supabase = create_client(url, key)


def normalizar(texto):
    return (
        str(texto or '')
        .encode('latin1', errors='ignore')
        .decode('utf-8', errors='ignore')
        .strip()
        .upper()
        .replace('Á', 'A')
        .replace('É', 'E')
        .replace('Í', 'I')
        .replace('Ó', 'O')
        .replace('Ú', 'U')
        .replace('Ã', 'A')
        .replace('Õ', 'O')
        .replace('Ç', 'C')
        .replace('Â', 'A')
        .replace('Ê', 'E')
        .replace('Ô', 'O')
        .replace('À', 'A')
    )


def parse_date(valor):
    if not valor or pd.isna(valor):
        return None
    s = str(valor).strip()
    m = re.match(r'(\d{1,2})/(\d{1,2})/(\d{4})', s)
    if not m:
        return None
    dia, mes, ano = m.groups()
    try:
        return f'{int(ano):04d}-{int(mes):02d}-{int(dia):02d}'
    except ValueError:
        return None


def main():
    print('=== Análise de Ocorrências Antigas ===\n')

    # 1. Lê planilha
    arquivo = Path('public/Ocorrências de advertência e suspensão para CORH em sem cpf 15jul26.xlsx')
    df = pd.read_excel(arquivo, dtype=str)
    df.columns = [normalizar(c) for c in df.columns]
    print('Colunas:', df.columns.tolist())
    print('Total de ocorrências na planilha:', len(df))

    # 2. Estatísticas básicas
    macros = df['MACRO'].dropna().astype(str).str.strip().unique().tolist()
    tipos = df['TIPO'].dropna().astype(str).str.strip().unique().tolist()
    nomes_unicos = df['FUNCIONARIO'].dropna().astype(str).str.strip().unique().tolist()

    print('Macro grupos:', macros)
    print('Tipos de ocorrência:', tipos)
    print('Funcionários únicos:', len(nomes_unicos))

    # 3. Valida datas
    df['DATA_OCORRIDO_PARSED'] = df['DATA DO OCORRIDO'].apply(parse_date)
    df['DATA_REGISTRO_PARSED'] = df['DATA DO REGISTRO'].apply(parse_date)
    datas_invalidas = df[(df['DATA_OCORRIDO_PARSED'].isna()) | (df['DATA_REGISTRO_PARSED'].isna())]
    print('Ocorrências com datas inválidas:', len(datas_invalidas))
    if len(datas_invalidas) > 0:
        print(datas_invalidas[['# DA OCORRENCIA', 'FUNCIONARIO', 'DATA DO OCORRIDO', 'DATA DO REGISTRO']].head().to_string(index=False))

    # 4. Busca colaboradores no banco
    print('\nBuscando colaboradores no banco...')
    colaboradores = []
    pagina = 0
    while True:
        resp = supabase.table('colaboradores').select('id, matricula, nome_completo, empresa_id, status').range(pagina * 1000, (pagina + 1) * 1000 - 1).execute()
        if not resp.data:
            break
        colaboradores.extend(resp.data)
        if len(resp.data) < 1000:
            break
        pagina += 1

    print('Total de colaboradores no banco:', len(colaboradores))

    # 5. Faz matching por nome
    mapa_nomes = {}
    for c in colaboradores:
        nome_norm = normalizar(c['nome_completo'])
        mapa_nomes.setdefault(nome_norm, []).append(c)

    encontrados = 0
    nao_encontrados = 0
    ambiguos = 0
    nomes_nao_encontrados = []

    for nome in nomes_unicos:
        nome_norm = normalizar(nome)
        matches = mapa_nomes.get(nome_norm)
        if not matches:
            nao_encontrados += 1
            nomes_nao_encontrados.append(nome)
        elif len(matches) > 1:
            ambiguos += 1
            print(f'Nome ambíguo ({len(matches)} colaboradores): {nome}')
        else:
            encontrados += 1

    print('\n=== Resultado do Matching por Nome ===')
    print('Funcionários encontrados (único):', encontrados)
    print('Funcionários ambíguos:', ambiguos)
    print('Funcionários não encontrados:', nao_encontrados)

    if nomes_nao_encontrados:
        print('\nExemplos de nomes não encontrados:')
        for nome in nomes_nao_encontrados[:30]:
            print('  -', nome)

    # 6. Mapeamento de tipos
    print('\n=== Mapeamento de Tipos ===')
    tipos_sistema = [
        'Advertência Verbal',
        'Advertência Escrita',
        'Suspensão 1 (1ª ocorrência)',
        'Suspensão 2 (reincidência)',
        'Suspensão 3 (3ª ocorrência)',
    ]
    for tipo_planilha in tipos:
        match = next((t for t in tipos_sistema if normalizar(t) == normalizar(tipo_planilha)), None)
        print(f'  "{tipo_planilha}" -> {match or "NÃO ENCONTRADO"}')


if __name__ == '__main__':
    main()

File: scripts/analisar-ocorrencias-antigas.mjs
import { createClient } from '@supabase/supabase-js'
import fs from 'fs'
import path from 'path'

function carregarEnv(caminho) {
  if (!fs.existsSync(caminho)) return
  const conteudo = fs.readFileSync(caminho, 'utf-8')
  for (const linha of conteudo.split('\n')) {
    const trimmed = linha.trim()
    if (!trimmed || trimmed.startsWith('#')) continue
    const idx = trimmed.indexOf('=')
    if (idx === -1) continue
    const chave = trimmed.slice(0, idx).trim()
    const valor = trimmed.slice(idx + 1).trim().replace(/^["']|["']$/g, '')
    process.env[chave] = valor
  }
}

carregarEnv(path.resolve(process.cwd(), '.env'))

const url = process.env.VITE_SUPABASE_URL
const key = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.VITE_SUPABASE_ANON_KEY

if (!url || !key) {
  console.error('Variáveis VITE_SUPABASE_URL e SUPABASE_SERVICE_ROLE_KEY são obrigatórias')
  process.exit(1)
}

const supabase = createClient(url, key)

function normalizar(str) {
  return String(str || '')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/[^a-zA-Z0-9\s]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
    .toUpperCase()
}

function parseCSV(caminho) {
  const conteudo = fs.readFileSync(caminho, 'utf-8')
  const linhas = conteudo.split(/\r?\n/).filter(l => l.trim())
  if (linhas.length === 0) return []
  const cabecalho = parseLinhaCSV(linhas[0])
  return linhas.slice(1).map(linha => {
    const valores = parseLinhaCSV(linha)
    const row = {}
    cabecalho.forEach((h, i) => {
      row[h] = valores[i] !== undefined ? valores[i] : ''
    })
    return row
  })
}

function parseLinhaCSV(linha) {
  const resultado = []
  let atual = ''
  let dentroAspas = false
  for (let i = 0; i < linha.length; i++) {
    const char = linha[i]
    const prox = linha[i + 1]
    if (char === '"') {
      if (dentroAspas && prox === '"') {
        atual += '"'
        i++
      } else {
        dentroAspas = !dentroAspas
      }
    } else if (char === ',' && !dentroAspas) {
      resultado.push(atual)
      atual = ''
    } else {
      atual += char
    }
  }
  resultado.push(atual)
  return resultado
}

function parseDate(valor) {
  if (!valor) return null
  const str = String(valor).trim()
  const partes = str.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/)
  if (!partes) return null
  const [, dia, mes, ano] = partes
  const data = new Date(Date.UTC(parseInt(ano, 10), parseInt(mes, 10) - 1, parseInt(dia, 10)))
  if (Number.isNaN(data.getTime())) return null
  return data.toISOString().split('T')[0]
}

async function main() {
  console.log('=== Análise de Ocorrências Antigas ===\n')

  // 1. Carrega CSV
  const arquivo = path.resolve(process.cwd(), 'public/ocorrencias_antigas.csv')
  const rows = parseCSV(arquivo)
  console.log('Total de ocorrências na planilha:', rows.length)

  // 2. Estatísticas básicas
  const macros = new Set()
  const tipos = new Set()
  const nomesPlanilha = new Set()
  const datasInvalidas = []

  for (const row of rows) {
    macros.add(String(row['macro'] || '').trim())
    tipos.add(String(row['tipo'] || '').trim())
    nomesPlanilha.add(String(row['funcionario'] || '').trim())
    const dataOcorrido = parseDate(row['data_do_ocorrido'])
    const dataRegistro = parseDate(row['data_do_registro'])
    if (!dataOcorrido || !dataRegistro) {
      datasInvalidas.push({
        nome: row['funcionario'],
        numero: row['#_da_ocorrencia'],
        dataOcorrido: row['data_do_ocorrido'],
        dataRegistro: row['data_do_registro'],
      })
    }
  }

  console.log('Macro grupos:', Array.from(macros))
  console.log('Tipos de ocorrência:', Array.from(tipos))
  console.log('Funcionários únicos:', nomesPlanilha.size)
  console.log('Ocorrências com datas inválidas:', datasInvalidas.length)
  if (datasInvalidas.length > 0) {
    console.log('Exemplos:', datasInvalidas.slice(0, 5))
  }

  // 3. Busca colaboradores no banco
  console.log('\nBuscando colaboradores no banco...')
  const colaboradores = []
  let pagina = 0
  while (true) {
    const { data, error } = await supabase
      .from('colaboradores')
      .select('id, matricula, nome_completo, empresa_id, status')
      .range(pagina * 1000, (pagina + 1) * 1000 - 1)
    if (error) {
      console.error('Erro ao buscar colaboradores:', error.message)
      process.exit(1)
    }
    if (!data || data.length === 0) break
    colaboradores.push(...data)
    if (data.length < 1000) break
    pagina++
  }
  console.log('Total de colaboradores no banco:', colaboradores.length)

  // 4. Faz matching por nome
  const mapaNomes = new Map()
  for (const c of colaboradores) {
    const nomeNormalizado = normalizar(c.nome_completo)
    if (!mapaNomes.has(nomeNormalizado)) {
      mapaNomes.set(nomeNormalizado, [])
    }
    mapaNomes.get(nomeNormalizado).push(c)
  }

  let encontrados = 0
  let ambiguos = 0
  let naoEncontrados = 0
  const nomesNaoEncontrados = []

  for (const nome of nomesPlanilha) {
    const nomeNormalizado = normalizar(nome)
    const matches = mapaNomes.get(nomeNormalizado)
    if (!matches) {
      naoEncontrados++
      nomesNaoEncontrados.push(nome)
    } else if (matches.length > 1) {
      ambiguos++
      console.log(`Nome ambíguo (${matches.length} colaboradores): ${nome}`)
    } else {
      encontrados++
    }
  }

  console.log('\n=== Resultado do Matching por Nome ===')
  console.log('Funcionários encontrados (único):', encontrados)
  console.log('Funcionários ambíguos:', ambiguos)
  console.log('Funcionários não encontrados:', naoEncontrados)

  if (nomesNaoEncontrados.length > 0) {
    console.log('\nExemplos de nomes não encontrados:')
    for (const nome of nomesNaoEncontrados.slice(0, 30)) {
      console.log('  -', nome)
    }
  }

  // 5. Mapeamento de tipos
  console.log('\n=== Mapeamento de Tipos ===')
  const tiposSistema = [
    'Advertência Verbal',
    'Advertência Escrita',
    'Suspensão 1 (1ª ocorrência)',
    'Suspensão 2 (reincidência)',
    'Suspensão 3 (3ª ocorrência)',
  ]
  for (const tipoPlanilha of Array.from(tipos)) {
    const match = tiposSistema.find(t => normalizar(t) === normalizar(tipoPlanilha))
    console.log(`  "${tipoPlanilha}" -> ${match || 'NÃO ENCONTRADO'}`)
  }
}

main().catch(err => {
  console.error(err)
  process.exit(1)
})
